graviton/blackbox.py

```diff
@@ -174,406 +196,385 @@ def _partial_encode_assert(
+        """
+        Validate and possibly return modified versions of:
+        * args
+        * abi_argument_types
+        """
+        args_out = list(args)
+        aats_out = copy(self.abi_argument_types)
+        nope: EncodingType = None
+        if self.validation:
+            assert aats_out is not None, "unexpected None"
+            assert self.method is not None, "unexpected None"
+            assert self.selector is not None, "unexpected None"
+            aats_out = cast(List[EncodingType], aats_out)
+            method = cast(abi.Method, self.method)
+            selector = cast(PyTypes, self.selector)
+            if len(args_out) == len(aats_out):
+                if not self.omit_method_selector:
+                    # the method selector is not abi-encoded, hence its abi-type is set to None
+                    aats_out = [None] + aats_out
+                    args_out = [selector] + args_out
+
+            elif len(args_out) == len(aats_out) + 1:
+                assert (
+                    args_out[0] == selector
+                ), f"{args_out[0]=} should have been the {selector=}"
+
+                if self.omit_method_selector:
+                    args_out = args_out[1:]
+                else:
+                    aats_out = [nope] + aats_out
+
+            else:
+                raise AssertionError(
+                    f"{len(args_out)=} is incompatible with {len(method.args)=}: LEFT should be equal or exactly RIGHT + 1"
+                )
+        elif not self.omit_method_selector:
+            assert aats_out is not None
+            aats_out = [nope] + cast(List[EncodingType], aats_out)
+        # else: not validating + omitting method selector ==> aats_out == abi_argument_types
+
+        return tuple(args_out), aats_out
+
+    @classmethod
+    def _prerun_validation(
+        cls, mode: ExecutionMode, abi_method_signature: Optional[str]
+    ) -> Tuple[
+        bool,
+        Optional[List[EncodingType]],
+        Optional[Union[abi.ABIType, str]],
+        Optional[abi.Method],
+        Optional[bytes],
+    ]:
+        assert (
+            len(ExecutionMode) == 2
+        ), f"assuming only 2 ExecutionMode's but have {len(ExecutionMode)}"
+        assert mode in ExecutionMode, f"unknown mode {mode} of type {type(mode)}"
+        is_app = mode == ExecutionMode.Application
+
+        abi_argument_types: Optional[List[EncodingType]] = None
+        abi_return_type: Optional[Union[abi.ABIType, str]] = None
+
+        method: Optional[abi.Method]
+        if abi_method_signature:
+            method = abi.Method.from_signature(abi_method_signature)
+            selector = method.get_selector()
+            abi_argument_types = [a.type for a in method.args]
+
+            if method.returns.type != abi.Returns.VOID:
+                abi_return_type = method.returns.type
+        else:  # logic sigs always land here:
+            method = selector = None
```

Comment:
I guess the previous `method = selector = None` suggestion also applies here.

Reply:
also keep the comment of logic sigs lands in the final case with no `method` and `selector`.

Reply:
see df452d53d4e5e1335708489b0dfb1ad9e807a12c

Reply:
lgtm